src/plugin_host.jl

```diff
@@ -0,0 +1,174 @@
+function on_notify(h::HostHandler, c, name::String, args::Vector{Any})
+    proc = require_plug(h, name)
+
+    if proc == nothing
+        throw(ErrorException(name * " not defined"))
+    end
+
+    @async(try
+        proc(c, args...)
+    catch
+        throw(MethodError(name, tuple(NvimClient, args...)))
+        throw(err)
+    end)
+end
+
+function on_request(h::HostHandler, c, serial, method, args)
+    if method == "specs" # called on UpdateRemotePlugins
+        reply_result(c, serial, get_specs(args...))
+    else
+        proc = require_plug(h, method)
+
+        if proc == nothing
+            reply_error(c, serial, "Command not defined.")
+        end
+
+        @async(try
+            reply_result(c, serial, proc(c, args...))
+        catch
+            err = "No matching method for args (NvimClient, " * join(args, ", ") * ")"
+            reply_error(c, serial, err)
+        end)
+    end
+end
+
+function require_plug(h::HostHandler, name::String)
+    (plugin_file, proc_id) = split(name, ':', 2)
+
+    if plugin_file ∉ h.loaded_plugins
+        cbs = get_callbacks(plugin_file)
+        merge!(h.proc_callbacks, cbs)
+    end
+
+    get(h.proc_callbacks, name, nothing)
+end
+
+function get_specs(plugin_file)
+    plugin = nothing
+    try
+        plugin = decorate(plugin_file)
+    end
+
+    fn_specs = {}
+    global plug(proc_type, name, handler, opt_args...) = begin
+        conf = Dict{String, Any}()
+        conf["type"] = proc_type
+        conf["name"] = name
+        opts = Dict{String, Any}()
+        for (opt_k, opt_v) in opt_args
+            opts[string(opt_k)] = opt_v
+        end
+        conf["opts"] = opts
+        conf["sync"] = pop!(opts, "sync", 0)
+        push!(fn_specs, conf)
+    end
+
+    try
+        eval(plugin)
+    catch err
+        println(STDERR, "Error while loading plugin " * plugin_file)
+        println(STDERR, err)
+    end
+
+    fn_specs
+end
+
+function get_callbacks(plugin_file)
+    plugin = nothing
+    try
+        plugin = decorate(plugin_file)
+    end
+
+    proc_callbacks = Array((String, Expr), 0)
+    global plug(proc_type, name, handler, opt_args...) = begin
+        pattern = ""
+        for arg in opt_args
+            if arg[1] == "pattern"
+                pattern = ":" * arg[2]
+            end
+        end
+
+        proc_name = "$plugin_file:$proc_type:$name$pattern"
+        push!(proc_callbacks, (proc_name, parse(handler)))
+    end
+
+    try
+        eval(plugin)
+    catch err
+        println(STDERR, "Error while loading plugin " * plugin_file)
+        println(STDERR, err)
+    end
+
+    extract_fns(x) = (x[1], eval(x[2]))
+    Dict{String, Function}(map(extract_fns, proc_callbacks))
+end
+
+# finds constructs that look like decorators (i.e. macrocall then function)
+# and then adds the function name as the last argument to the macrocall
+function decorate(file_name::String)
+    ast = (open(readall, file_name) |> parse)::Expr
```

Comment:
Not sure I like this solution. We should be able just require the plugin module as any julia module (i e it will be executed at most once) and the "decorators" should just register all information (both specs and the func reference) at once.
I could imagine the "cleanest" solution to have some module level metadata `MyPlugin._nvim_handlers` but maybe it is simpler to just have a `global host = HostHandler()` and add a `spec::Dict{ByteString,Dict}` and just let `plug()` register into it. I will look into it more later.

(Later on, when julia supports modular precomilation, I like to imagine supporting to have a precompiled host module that already has required all the plugin modules, and the metadata already precomputed)